pkg/filter/manager.go

```diff
@@ -0,0 +1,276 @@
+// Package filter implements the filter extension of mosn
+package filter
+
+import (
+	"context"
+
+	"mosn.io/api"
+	"mosn.io/mosn/pkg/types"
+	"mosn.io/pkg/buffer"
+)
+
+// UndefinedFilterPhase undefined filter phase, used for senderFilter.
+const UndefinedFilterPhase api.FilterPhase = 99999
+
+// StreamFilterChainStatus determines the running status of filter chain.
+type StreamFilterChainStatus int
+
+const (
+	// StreamFilterChainContinue continues running the filter chain.
+	StreamFilterChainContinue StreamFilterChainStatus = 0
+	// StreamFilterChainStop stops running the filter chain, next time should retry the current filter.
+	StreamFilterChainStop StreamFilterChainStatus = 1
+	// StreamFilterChainReset stops running the filter chain and reset index, next time should run the first filter.
+	StreamFilterChainReset StreamFilterChainStatus = 2
+)
+
+// StreamFilterStatusHandler converts api.StreamFilterStatus to StreamFilterChainStatus.
+type StreamFilterStatusHandler func(status api.StreamFilterStatus) StreamFilterChainStatus
+
+// DefaultStreamFilterStatusHandler is the default implementation of StreamFilterStatusHandler.
+func DefaultStreamFilterStatusHandler(status api.StreamFilterStatus) StreamFilterChainStatus {
+	switch status {
+	case api.StreamFilterContinue:
+		return StreamFilterChainContinue
+	case api.StreamFilterStop:
+		return StreamFilterChainReset
```

Comment:
By default, every time we call RunXXXFilter should start from index=0, instead of starting from the last position.
Otherwise some filters may not be executed since the last call to RunXXXFilter skip them because of their mismatched Phase